### ISSUE TITLE
buffer.hpp broken with libc++

### ISSUE THREAD

**Issue by nazavode** [illuhad/hipSYCL#243]
Current `libc++` support for `std::reinterpret_pointer_cast` [is still missing](https://reviews.llvm.org/D70754), this makes any attempt to build SYCL sources with `-stdlib=libc++` to fail:

```console
$ syclcc -stdlib=libc++ -std=c++17 sycl.cpp
In file included from sycl.cpp:4:
In file included from /hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:47:
/hipSYCL/bin/../include/CL/../hipSYCL/sycl/buffer.hpp:365:36: error: no member named
      'reinterpret_pointer_cast' in namespace 'std'
    this->_writeback_buffer = std::reinterpret_pointer_cast<T>(other._writeback_buffer);
```

**Comment by illuhad** [comment]
hm.. hipSYCL now requires C++17. `std::reinterpret_pointer_cast` is part of C++17. So, I guess we should just wait for `libc++` to support that..

**Comment by nazavode** [comment]
Yep that makes sense, I was really disappointed to discover that `libc++` still lacks some C++17 stuff... Anyway, do you see something like a back port into hipSYCL (it's a matter of 4 lines of code: https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast) as a no no?

**Comment by fodinabor** [comment]
one could surely add a shim header, that defines the function if necessary and otherwise just import it into a hipSYCL namespace...

**Comment by nazavode** [comment]
@fodinabor exactly, if @illuhad says fine I'm obviously willing to provide a PR :)

**Comment by illuhad** [comment]
@nazavode If you file a PR I will merge it to improve the situation on Mac which is already non-optimal, but  don't want to spend much time on this myself as the SYCL world is now rapidly moving to C++17.

**Comment by nazavode** [comment]
@illuhad that makes perfect sense, and any `+` in a diff is a line that must be maintained. I will provide a PR but obviously feel free to reject it!

**Comment by arkantos493** [comment]
I've added a PR since I had the same problem after installing hipSYCL via spack.

**Comment by illuhad** [comment]
Fixed by PR #257. Thanks, @arkantos493 !

**Comment by nazavode** [comment]
Thank you so much @arkantos493 for having this one fixed!
